File: src/spatial_operators/gradient_operators.f90
! This source file is part of the Geophysical Fluids Modeling Framework (GAME), which is released under the MIT license.
! Github repository: https://github.com/OpenNWP/GAME

module gradient_operators

  ! This module contains the gradient operators.
  
  use iso_c_binding
  use definitions, only: wp
  use grid_nml,    only: n_vectors,n_vectors_h,n_layers,n_scalars,n_scalars_h,n_vectors_per_layer, &
                         n_v_vectors
  
  implicit none
  
  private
  
  public :: grad_hor_cov
  public :: grad_vert_cov
  
  contains

  subroutine grad_hor_cov(in_field,out_field,from_index,to_index,normal_distance) &
  bind(c,name = "grad_hor_cov")
  
    ! This subroutine calculates the horizontal covariant gradient
    
    real(wp),       intent(in)  :: in_field(n_scalars)
    real(wp),       intent(out) :: out_field(n_vectors)
    integer(c_int), intent(in)  :: from_index(n_vectors_h)
    integer(c_int), intent(in)  :: to_index(n_vectors_h)
    real(wp),       intent(in)  :: normal_distance(n_vectors)
    
    ! local variables
    integer :: h_index,layer_index,vector_index
    
    !$omp parallel do private(h_index,layer_index,vector_index)
    do h_index=1,n_vectors_h
      do layer_index=0,n_layers-1
        vector_index = n_scalars_h + layer_index*n_vectors_per_layer + h_index
        out_field(vector_index) &
        = (in_field(1+to_index(h_index)+layer_index*n_scalars_h) &
        - in_field(1+from_index(h_index)+layer_index*n_scalars_h)) &
        /normal_distance(vector_index)
      enddo
    enddo
    !$omp end parallel do
    
  end subroutine grad_hor_cov
  
  subroutine grad_vert_cov(in_field,out_field,normal_distance) &
  bind(c,name = "grad_vert_cov")
  
    ! This subroutine calculates the vertical covariant gradient.
    
    real(wp), intent(in)  :: in_field(n_scalars)
    real(wp), intent(out) :: out_field(n_vectors)
    real(wp), intent(in)  :: normal_distance(n_vectors)
    
    integer :: ji,layer_index,h_index,lower_index,upper_index,vector_index
    
    ! loop over the inner grid points
    !$omp parallel do private(ji,layer_index,h_index,lower_index,upper_index,vector_index)
    do ji=n_scalars_h+1,n_v_vectors-n_scalars_h
      layer_index = (ji-1)/n_scalars_h
      h_index = ji - layer_index*n_scalars_h
      lower_index = h_index + layer_index*n_scalars_h
      upper_index = h_index + (layer_index-1)*n_scalars_h
      vector_index = h_index + layer_index*n_vectors_per_layer
      out_field(vector_index) &
      = (in_field(upper_index)-in_field(lower_index))/normal_distance(vector_index)
    enddo
    !$omp end parallel do
  
  end subroutine grad_vert_cov

end module gradient_operators

File: src/spatial_operators/gradient_operators.c
/*
This source file is part of the Geophysical Fluids Modeling Framework (GAME), which is released under the MIT license.
Github repository: https://github.com/OpenNWP/GAME
*/

/*
This file contains the gradient operators.
*/

#include <stdio.h>
#include "../game_types.h"
#include "spatial_operators.h"

int grad(Scalar_field in_field, Vector_field out_field, Grid *grid)
{
	/*
	calculates the gradient (horizontally contravariant, vertically covariant)
	*/
	grad_cov(in_field, out_field, grid -> from_index, grid -> to_index, grid -> normal_distance);
	vector_field_hor_cov_to_con(out_field, grid -> from_index, grid -> to_index, grid -> inner_product_weights, grid -> slope);
    return 0;
}

int grad_hor(Scalar_field in_field, Vector_field out_field, Grid *grid)
{
	/*
	This function calculates the horizontal contravariant gradient.
	*/
	grad(in_field, out_field, grid);
    int layer_index, h_index;
	#pragma omp parallel for private(layer_index, h_index)
    for (int i = 0; i < N_V_VECTORS; ++i)
    {
        layer_index = i/N_SCALS_H;
        h_index = i - layer_index*N_SCALS_H;
        out_field[h_index + layer_index*N_VECS_PER_LAYER] = 0;
    }
    return 0;
}




